Products/CMFPlone/recyclebin.py

```diff
@@ -0,0 +1,584 @@
+from datetime import datetime
+from datetime import timedelta
+from persistent.mapping import PersistentMapping
+from plone.registry.interfaces import IRegistry
+from Products.CMFPlone.controlpanel.browser.recyclebin import (
+    IRecycleBinControlPanelSettings,
+)
+from Products.CMFPlone.interfaces.recyclebin import IRecycleBin
+from zope.annotation.interfaces import IAnnotations
+from zope.component import getUtility
+from zope.component.hooks import getSite
+from zope.interface import implementer
+
+import logging
+import uuid
+
+
+logger = logging.getLogger("Products.CMFPlone.RecycleBin")
+
+ANNOTATION_KEY = "Products.CMFPlone.RecycleBin"
+
+
+@implementer(IRecycleBin)
+class RecycleBin:
+    """Stores deleted content items"""
+
+    def __init__(self, context=None):
+        """Initialize the recycle bin for a site
+
+        Args:
+            context: The Plone site object (optional when used as a utility)
+        """
+        self.context = context
+        # When used as a utility without context, we'll get the context on demand
+
+    def _get_context(self):
+        """Get the context (Plone site) if not already available"""
+        if self.context is None:
+            self.context = getSite()
+        return self.context
+
+    def _get_storage(self):
+        """Get the storage for recycled items"""
+        context = self._get_context()
+        annotations = IAnnotations(context)
+
+        if ANNOTATION_KEY not in annotations:
+            annotations[ANNOTATION_KEY] = PersistentMapping()
+
+        return annotations[ANNOTATION_KEY]
+
+    # Update property for storage to use _get_storage
+    @property
+    def storage(self):
+        return self._get_storage()
+
+    def _get_settings(self):
+        """Get recycle bin settings from registry"""
+        registry = getUtility(IRegistry)
+        return registry.forInterface(
+            IRecycleBinControlPanelSettings, prefix="plone-recyclebin"
+        )
+
+    def is_enabled(self):
+        """Check if recycle bin is enabled"""
+        try:
+            settings = self._get_settings()
+            return settings.recycling_enabled
+        except (KeyError, AttributeError):
+            return False
+
+    def add_item(self, obj, original_container, original_path, item_type=None):
+        """Add deleted item to recycle bin"""
+        if not self.is_enabled():
+            return None
+
+        # Generate a unique ID for the recycled item
+        item_id = str(uuid.uuid4())
```

Comment:
The object already has a UID; why not use it?

Reply:
i generated a new ID during addition as, after the object being deleted that same ID might be assigned to some other element and while the restoring it can create some conflicts?!

Reply:
Plone UIDs are also generated the same way, using uuid.uuid4(). There is a 1 in 2^128 chance that a new UID duplicates an existing one, and usually we assume it will not happen in practice.

Reply:
so nothing to be done here? @davisagli